### ISSUE TITLE
[BUG] rmm_allocator::deallocate(): RMM_FREE: initialization error

### ISSUE THREAD

**Issue by davidwendt** [rapidsai/rmm#85]
For Python classes that wrap C++ classes that contain memory allocated by RMM, when the python process ends the Python order of de-allocation may cause an RMM_FREE initialization error when using the pool allocator. This occurs because RMM instance may have been destroyed before the Python class. This error causes the python process to terminate with a core dump instead of cleanly exiting.

Simple testcase to show the error from a python command-line interpretter:
```
>>> from librmm_cffi import librmm as rmm
>>> from librmm_cffi import librmm_config as rmm_cfg
>>> rmm_cfg.use_pool_allocator = True 
>>> rmm.initialize()
0
>>> import nvstrings
>>> strs = nvstrings.to_device(["hello"])
>>> exit()
```
Before the process ends cleanly the following exception occurs terminating the process:
```
terminate called after throwing an instance of 'thrust::system::system_error'
  what():  rmm_allocator::deallocate(): RMM_FREE: cudaErrorInitializationError: initialization error
Aborted
```
This particular error is thrown in https://github.com/rapidsai/rmm/blob/919fb5b659d197435163384e4745b42bbe6ae3b8/include/rmm/thrust_rmm_allocator.h#L59
```
    inline void deallocate(pointer ptr, size_t)
    {
      rmmError_t error = RMM_FREE(thrust::raw_pointer_cast(ptr), stream);
  
      if(error != RMM_SUCCESS)
      {
        throw thrust::system_error(error, thrust::cuda_category(), "rmm_allocator::deallocate(): RMM_FREE");
      }
    }
```
The nvstrings instance points to a C++ NVStrings instance which has a member variable allocated with rmm::device_vector and this vector is freed after RMM is deinitialized by the python process.

Throwing the error in the rmm::device_vector destructor causes the process to terminator (core dump).

Propose checking for this condition (free called after deinit) inside of `RMM_FREE` or `rmm::free` and ignoring this error since the memory has already been freed and no corruption will occur.

**Comment by jrhemstad** [comment]
This sounds like Python is "doing the wrong thing".

Doing this properly and in a thread-safe manner within RMM would likely require adding a mutex to `rmmFinalize` and `rmm::free`, as well as adding additional checks to ensure that the memory manager is still initialized before attempting to free.  

My feeling is that a memory allocator should be as simple and lightweight as possible, and the onus is on the user to do the right thing.

Before we add additional complexity to `rmm:free` (read, make it slower), I'd hope we can explore if it's possible on the Python side to avoid finalizing the memory manager before the rest of the objects have been cleaned up. CC @kkraus14

**Comment by nouiz** [comment]
We had this problem in another software. In the past, it was a Python
restriction. Object aren't guaranteed to be destroyed.

The work around we used is to record when the "master" object is destroyed.
And make later function on it no-op. This could just add an if condition in
free().
Do you think something like this would be possible in this context too?

**Comment by davidwendt** [comment]
In this case the 'master' object is the `rmm:::device_vector`. The vector object itself lives in CPU memory and manages a pointer array (GPU memory). So the `deallocate` method mentioned above would need this logic.

**Comment by kkraus14** [comment]
Fixed by #149
